rpcs3/Emu/RSX/rsx_utils.h

```diff
@@ -200,7 +202,478 @@ namespace rsx
+		}
+
+		inline bool inside(const address_range_vector &vec) const;
+		inline bool overlaps(const address_range_vector &vec) const;
+
+		inline bool touches(const address_range &other) const
+		{
+			AUDIT(valid() && other.valid());
+			// returns true if there is overlap, or if sections are side-by-side
+			return overlaps(other) || other.start == next_address() || other.end == prev_address();
+		}
+
+		inline s32 signed_distance(const address_range &other) const
+		{
+			if (touches(other))
+				return 0;
+
+			// other after this
+			if (other.start > end)
+				return (s32)(other.start - end - 1);
+
+			// this after other
+			AUDIT(start > other.end);
+			return -((s32)(start - other.end - 1));
+		}
+
+		inline u32 distance(const address_range &other) const
+		{
+			if (touches(other))
+				return 0;
+
+			// other after this
+			if (other.start > end)
+				return (other.start - end - 1);
+
+			// this after other
+			AUDIT(start > other.end);
+			return (start - other.end - 1);
+		}
+
+		// Utilities
+		inline address_range get_min_max(const address_range &other) const
+		{
+			return {
+				std::min(valid() ? start : UINT32_MAX, other.valid() ? other.start : UINT32_MAX),
+				std::max(valid() ? end : 0, other.valid() ? other.end : 0)
+			};
+		}
+
+		void set_min_max(const address_range &other)
+		{
+			const address_range _range = get_min_max(other);
+			start = _range.start;
+			end = _range.end;
+		}
+
+		inline bool is_page_range() const
+		{
+			return (valid() && start % 4096u == 0 && length() % 4096u == 0);
+		}
+
+		inline address_range to_page_range() const
+		{
+			AUDIT(valid());
+			return { page_start(start), page_end(end) };
+		}
+
+		inline void page_align()
+		{
+			AUDIT(valid());
+			start = page_start(start);
+			end = page_end(end);
+			AUDIT(is_page_range());
+		}
+
+		inline address_range get_intersect(const address_range &clamp) const
+		{
+			if (!valid() || !clamp.valid())
+				return {};
+
+			return { std::max(start, clamp.start), std::min(end, clamp.end) };
+		}
+
+		inline void intersect(const address_range &clamp)
+		{
+			if (!clamp.valid())
+			{
+				invalidate();
+			}
+			else
+			{
+				start = std::max(start, clamp.start);
+				end = std::min(end, clamp.end);
+			}
+		}
+
+		// Validity
+		inline bool valid() const
+		{
+			return (start <= end);
+		}
+
+		inline void invalidate()
+		{
+			start = UINT32_MAX;
+			end = 0;
+		}
+
+		// Comparison Operators
+		inline bool operator ==(const address_range &other) const
+		{
+			return (start == other.start && end == other.end);
+		}
+
+		inline bool operator !=(const address_range &other) const
+		{
+			return (start != other.start || end != other.end);
+		}
+
+		/**
+		 * Debug
+		 */
+		inline std::string str() const
+		{
+			return fmt::format("{0x%x->0x%x}", start, end);
+		}
+	};
+
+	static inline address_range page_for(u32 addr)
+	{
+		return address_range::create_start_end(page_start(addr), page_end(addr));
+	}
+
+	class address_range_vector
```

Comment:
I don't like the idea of wrapping a vector in a class and then repeat its traits. Necessary functions can be implemented as free-standing functions.

Reply:
Honestly makes the code much cleaner since it encapsulates the merge/exclude/overlap/etc methods very similarly to the address_range class. This means I don't need to remember what the method was called (was it "address_range_merge_range_into_vector(vector, range);"? "merge_address_range_range_into_vector(vector, range)"? or maybe even "merge_range_into_address_vector(vector,range)"?).

I don't do this sort of encapsulation often, but in this case I think it's better than the alternatives.